File: frontend/src/lib/simulator.ts
import { Spot, SpotStatus } from '@/types';

// Time-of-day occupancy targets (Dubai local time UTC+4)
const OCCUPANCY_PROFILES = {
  night: { hours: [0, 6] as [number, number], range: [0.08, 0.15] as [number, number] },
  morningRush: { hours: [7, 10] as [number, number], range: [0.45, 0.85] as [number, number] },
  midMorning: { hours: [10, 12] as [number, number], range: [0.70, 0.80] as [number, number] },
  lunch: { hours: [12, 13] as [number, number], range: [0.60, 0.70] as [number, number] },
  afternoon: { hours: [13, 17] as [number, number], range: [0.72, 0.82] as [number, number] },
  eveningRush: { hours: [17, 20] as [number, number], range: [0.85, 0.92] as [number, number] },
  evening: { hours: [20, 24] as [number, number], range: [0.30, 0.50] as [number, number] },
};

export function getTargetOccupancy(hour: number): number {
  for (const profile of Object.values(OCCUPANCY_PROFILES)) {
    const [start, end] = profile.hours;
    if (hour >= start && hour < end) {
      const [min, max] = profile.range;
      return min + Math.random() * (max - min);
    }
  }
  // Fallback (hour 6-7 gap)
  return 0.3;
}

export function simulateTick(spots: Spot[], targetOccupancy: number): Spot[] {
  const totalSpots = spots.length;
  const currentOccupied = spots.filter(s => s.status === 'occupied' || s.status === 'reserved').length;
  const currentOccupancy = currentOccupied / totalSpots;

  const diff = targetOccupancy - currentOccupancy;
  const spotsToFlip = Math.max(1, Math.min(3, Math.ceil(Math.abs(diff) * totalSpots * 0.1)));

  const newSpots = [...spots];

  for (let i = 0; i < spotsToFlip; i++) {
    const idx = Math.floor(Math.random() * totalSpots);
    const spot = newSpots[idx];

    // Skip sensor_offline spots
    if (spot.status === 'sensor_offline') continue;

    let newStatus: SpotStatus;

    if (diff > 0) {
      // Need more occupancy — flip free spots to occupied
      if (spot.status === 'free') {
        newStatus = Math.random() < 0.85 ? 'occupied' : 'reserved';
      } else {
        continue;
      }
    } else {
      // Need less occupancy — flip occupied/reserved to free
      if (spot.status === 'occupied' || spot.status === 'reserved') {
        newStatus = 'free';
      } else {
        continue;
      }
    }

    newSpots[idx] = {
      ...spot,
      status: newStatus,
      last_changed_at: new Date().toISOString(),
      ...(newStatus === 'occupied'
        ? { occupied_since: new Date().toISOString() }
        : { occupied_since: undefined }),
    };
  }

  return newSpots;
}

File: backend/simulator.py
import asyncio
import logging
import random
from typing import Tuple
from datetime import datetime, timezone, timedelta
from sqlalchemy import select, update
from database import async_session
from models import Spot, Zone

logger = logging.getLogger(__name__)

# Dubai is UTC+4
DUBAI_OFFSET = timedelta(hours=4)

# Time-of-day occupancy profiles (Dubai local time)
PROFILES = {
    (0, 6): (0.08, 0.15),     # Night: 8-15%
    (7, 10): (0.45, 0.85),    # Morning rush: ramp 45→85%
    (10, 12): (0.70, 0.80),   # Mid-morning: stable 70-80%
    (12, 13): (0.60, 0.70),   # Lunch dip: 60-70%
    (13, 17): (0.72, 0.82),   # Afternoon: 72-82%
    (17, 20): (0.85, 0.92),   # Evening rush: 85-92%
    (20, 24): (0.30, 0.50),   # Evening: winding down
}


def get_target_occupancy() -> Tuple[float, float]:
    """Get target occupancy range based on current Dubai time."""
    now_utc = datetime.now(timezone.utc)
    dubai_time = now_utc + DUBAI_OFFSET
    hour = dubai_time.hour

    for (start, end), (low, high) in PROFILES.items():
        if start <= hour < end:
            return (low, high)
    return (0.08, 0.15)  # fallback to night


async def simulate_tick(changed_spots: list):
    """Run one simulation tick: adjust spot statuses toward target occupancy."""
    async with async_session() as session:
        # Get all zones with spots
        result = await session.execute(select(Zone))
        zones = result.scalars().all()

        target_low, target_high = get_target_occupancy()
        target_mid = (target_low + target_high) / 2

        for zone in zones:
            spots = zone.spots
            if not spots:
                continue

            total = len(spots)
            occupied = sum(1 for s in spots if s.status == "occupied")
            current_occ = occupied / total if total > 0 else 0

            # Calculate how far we are from target
            diff = target_mid - current_occ

            for spot in spots:
                if spot.status in ("reserved", "sensor_offline"):
                    continue

                # Probability of flip proportional to distance from target
                flip_prob = min(abs(diff) * 0.3, 0.15)

                if random.random() < flip_prob:
                    now = datetime.now(timezone.utc)
                    if diff > 0 and spot.status == "free":
                        # Need more occupied spots
                        spot.status = "occupied"
                        spot.occupied_since = now
                        spot.last_changed_at = now
                        changed_spots.append({
                            "id": spot.id,
                            "status": "occupied",
                            "last_changed_at": now.isoformat()
                        })
                    elif diff < 0 and spot.status == "occupied":
                        # Need more free spots
                        spot.status = "free"
                        spot.occupied_since = None
                        spot.last_changed_at = now
                        changed_spots.append({
                            "id": spot.id,
                            "status": "free",
                            "last_changed_at": now.isoformat()
                        })

        await session.commit()


async def run_simulator(broadcast_fn):
    """Main simulator loop - runs every 2 seconds."""
    while True:
        try:
            changed_spots = []
            await simulate_tick(changed_spots)
            if changed_spots and broadcast_fn:
                await broadcast_fn({
                    "type": "spot_update",
                    "spots": changed_spots
                })
        except Exception as e:
            logger.error(f"Simulator error: {e}")
        await asyncio.sleep(2)
